direct-messages/README.md

```diff
@@ -10,24 +10,74 @@ because the same message can simultaneously:
+
+  return hkdf(input_keying_material, length, {salt, info, hash})
 }
 ```
 
-Notes:
-- we (curently) use the primary feed keys for this derivation
-- for feeds based on `ed25519` keypairs, we convert these to `curve25519` keypairs before doing scalarmult
-
+The function arguments are:
+- `my_dh_secret`: the own Diffie-Hellman secret key.
+- `my_dh_public`: the own Diffie-Hellman public key.
+- `your_dh_public`: the Diffie-Hellman public key of the communication partner.
+- `my_feed_tfk`: the own Scuttlebutt feed ID in tfk format, as specified in TODO.
+- `your_feed_tfk`: the Scuttlebutt feed ID in tfk format of the communication partner.
+
+The constants we use are:
+- `hash` is set to `SHA256`, such that HKDF uses SHA256 as underlying hash function.
+- `length` is set to 32, such that the output keys are 32 bytes = 256 bits long.
+- `salt` is set to `SHA256("envelope-dm-v1-extract-salt")`.
+
+Furthermore, we require the following helper functions:
+- `SHA256`: The SHA256 hash function
+- `slp_encode`: Returns the shallow length prefix encoding of a tuple/list, [as defined in the envelope spec](https://github.com/ssbc/envelope-spec/blob/master/encoding/slp.md).
+- `sort`: Sort an array. If the array elements are buffers, sort the elements such that any buffers `b1`, `b2` with leading bytes `b1[i] < b2[i]` are ordered such that `b1` appears before `b2`.
+- `scalarmult`: for example, the curve25519 scalar multiplication function. As common in elliptic curve cryptography, we use additive group notation. Should the underlying Diffie-Hellman algorithms be exchanged, the call is substituted with the respective scalar multiplication or exponentiation function.
+- `hkdf`: The HKDF function, as specified in [RFC 5869]
+- `tfk_encode`: The binary encoding of keys, as specified in TODO
+
+As a high level overview, we first compute the shared key between our secret key and the other public key.
+Then, we derive a secret from it using HKDF.
+The key derivation function takes four inputs: the input keying material, the key length, a salt and some additional info that ties the output key to the context.
+A different info value will make the key derivation return a different value, even if the same input key is used.
+This means the `info` variable can be used to tie the output key to the specific context of this call. 
+That context is composed of several values.
+
+Firstly, it contains a purpose identifying this particular key derivation.
+
+Secondly, it contains the two Diffie-Hellman public keys whose shared secret is calculated, concatenated with the respective feed ID.
+  We include the Diffie-Hellman public keys, because it is possible for two pairs `(skA, pkB), (skC, pkD)` to have scalar multiplication result in the same shared key `skA*pkB=skC*pkD`.
+  By deriving based on the two public keys, we disambiguate which of the possible keys were used and return a different output key, respectively.
```

Comment:
> which of the possible keys
:fire: in this context, the word "keys" alone is dangerous because there are so many keys. i.e. can you say which keys your're talking about, because I am not sure.

> output key
similar, not sure which output key you mean. If you mean the final `shared_dm_key` then say that. Overall didn't know what this sentence was pointing at becaue I'm a crypto noob and the language was flexi <3